x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/mapper/SemanticTextInferenceResultFieldMapper.java

```diff
@@ -0,0 +1,241 @@
+        }
+
+        parseFieldInferenceResultsArray(context, mapperBuilderContext, fieldName);
+    }
+
+    private static void parseFieldInferenceResultsArray(
+        DocumentParserContext context,
+        MapperBuilderContext mapperBuilderContext,
+        String fieldName) throws IOException {
+
+        XContentParser parser = context.parser();
+        NestedObjectMapper nestedObjectMapper = createNestedObjectMapper(context, mapperBuilderContext, fieldName);
+        context.path().add(fieldName); // TODO: Need path manipulation?
+
+        try {
+            if (parser.nextToken() != XContentParser.Token.START_ARRAY) {
+                throw new DocumentParsingException(parser.getTokenLocation(), "Expected a START_ARRAY, got " + parser.currentToken());
+            }
+
+            for (XContentParser.Token token = parser.nextToken(); token != XContentParser.Token.END_ARRAY; token = parser.nextToken()) {
+                DocumentParserContext nestedContext = context
+                    .createChildContext(nestedObjectMapper)
+                    .createNestedContext(nestedObjectMapper);
+
+                if (token != XContentParser.Token.START_OBJECT) {
+                    throw new DocumentParsingException(parser.getTokenLocation(), "Expected a START_OBJECT, got " + parser.currentToken());
+                }
+
+                for (token = parser.nextToken(); token != XContentParser.Token.END_OBJECT; token = parser.nextToken()) {
+                    if (token != XContentParser.Token.FIELD_NAME) {
+                        throw new DocumentParsingException(parser.getTokenLocation(),
+                            "Expected a FIELD_NAME, got " + parser.currentToken());
+                    }
+
+                    String currentName = parser.currentName();
+                    context.path().add(currentName);
+                    try {
+                        // TODO: Test how this code handles extra/missing fields
+                        FieldMapper childNestedMapper = (FieldMapper) nestedObjectMapper.getMapper(currentName);
+                        if (childNestedMapper == null) {
+                            throw new DocumentParsingException(parser.getTokenLocation(), "Unexpected field name: " + currentName);
+                        }
+                        parser.nextToken();
+                        childNestedMapper.parse(nestedContext);
+                    } finally {
+                        context.path().remove();
+                    }
+                }
+            }
+        } finally {
+            context.path().remove();
+        }
+    }
+
+    private static NestedObjectMapper createNestedObjectMapper(
+        DocumentParserContext context,
+        MapperBuilderContext mapperBuilderContext,
+        String fieldName) {
+
+        // TODO: Use keyword field type for text?
+        // TODO: Why add text field to mapper if it's not indexed or stored?
```

Comment:
@carlosdelest Basically what the TODO says. Why add `text` to the mapping if it's not stored or indexed?

Reply:
The idea was to validate the text but not doing anything about it. It's basically a sanity check that we're receiving the text field and it's valid, so we ensure we can retrieve that from source later. I'm totally open for discussion on what to do that doesn't imply indexing data.

Reply:
>         // TODO: Why add text field to mapper if it's not indexed or stored?

We shouldn't it needs to be in _source only.

> It's basically a sanity check that we're receiving the text field and it's valid, so we ensure we can retrieve that from source later. 

The upstream inference action should be doing this right? We will get inference failures if its not text.